providers/asrockrack/firmware.go

```diff
@@ -21,125 +22,167 @@ const (
-	switch component {
+	return nil, errors.Wrap(bmclibErrs.ErrFirmwareUpload, "component unsupported: "+component)
+}
+
+func (a *ASRockRack) FirmwareUpload(ctx context.Context, component string, file *os.File) (taskID string, err error) {
+	switch strings.ToUpper(component) {
 	case common.SlugBIOS:
-		err = a.firmwareInstallBIOS(ctx, reader, size)
+		return "", a.firmwareUploadBIOS(ctx, file)
 	case common.SlugBMC:
-		err = a.firmwareInstallBMC(ctx, reader, size)
-	default:
-		return "", errors.Wrap(bmclibErrs.ErrFirmwareInstall, "component unsupported: "+component)
+		return "", a.firmwareUploadBMC(ctx, file)
 	}
 
-	if err != nil {
-		err = errors.Wrap(bmclibErrs.ErrFirmwareInstall, err.Error())
-	}
+	return "", errors.Wrap(bmclibErrs.ErrFirmwareUpload, "component unsupported: "+component)
 
-	return jobID, err
 }
 
-// FirmwareInstallStatus returns the status of the firmware install process, a bool value indicating if the component requires a reset
-func (a *ASRockRack) FirmwareInstallStatus(ctx context.Context, installVersion, component, taskID string) (status string, err error) {
-	component = strings.ToUpper(component)
-	switch component {
-	case common.SlugBIOS, common.SlugBMC:
-		return a.firmwareUpdateStatus(ctx, component, installVersion)
-	default:
-		return "", errors.Wrap(bmclibErrs.ErrFirmwareInstallStatus, "component unsupported: "+component)
+func (a *ASRockRack) firmwareUploadBMC(ctx context.Context, file *os.File) error {
+	//	// expect atleast 5 minutes left in the deadline to proceed with the upload
+	d, _ := ctx.Deadline()
+	if time.Until(d) < 5*time.Minute {
+		return errors.New("remaining context deadline insufficient to perform update: " + time.Until(d).String())
 	}
-}
-
-// firmwareInstallBMC uploads and installs firmware for the BMC component
-func (a *ASRockRack) firmwareInstallBMC(ctx context.Context, reader io.Reader, fileSize int64) error {
-	var err error
 
-	// 0. take the model so that we use a different endpoint on E3C256D4ID-NL
-	device := common.NewDevice()
-	device.Metadata = map[string]string{}
-	err = a.fruAttributes(ctx, &device)
-	if err != nil {
-		return errors.Wrap(err, "failed to get model in step 0/4")
-	}
-
-	// 1. set the device to flash mode - prepares the flash
 	// Beware: this locks some capabilities, e.g. the access to fruAttributes
 	a.log.V(2).WithValues("step", "1/4").Info("set device to flash mode, takes a minute...")
-	err = a.setFlashMode(ctx)
+	err := a.setFlashMode(ctx)
 	if err != nil {
-		return errors.Wrap(err, "failed in step 1/4 - set device to flash mode")
+		return errors.Wrap(
+			bmclibErrs.ErrFirmwareUpload,
+			"failed in step 1/3 - set device to flash mode: "+err.Error(),
+		)
 	}
 
-	// 2. upload firmware image file
-	fwEndpoint := "api/maintenance/firmware"
+	var fwEndpoint string
+	switch a.deviceModel {
 	// E3C256D4ID-NL calls a different endpoint for firmware upload
-	if strings.EqualFold(device.Model, "E3C256D4ID-NL") {
+	case "E3C256D4ID-NL":
```

Comment:
This should also probably be a constant. The same one you create for the string [here](https://github.com/bmc-toolbox/bmclib/pull/378/files#diff-95ef4ebc48f6ac687634616a33c0934fe7ebf83738e6de6e10ca761db069a391R122).